### PR TITLE
BD-2550. Link tokens

### PR DESCRIPTION
**Pull request by trukhilio**

### CODE REVIEW

File: client/src/constants/common.ts [hunk 1]
```diff
@@ -29,6 +29,7 @@ export const CONFIRM_PIN_ACTIONS = {
   ENABLE_TWO_FACTOR: 'ENABLE_TWO_FACTOR',
   TWO_FACTOR_REQUEST: 'TWO_FACTOR_REQUEST',
   SHOW_BACKUP_CODE: 'SHOW_BACKUP_CODE',
+  TOKEN_LIST: 'TOKEN_LIST',
```
Comment (andreyvEze) [inline client/src/constants/common.ts:32]: let's create const for each action - add/edit/delete

File: client/src/pages/AddTokenPage/AddTokenPage.tsx [hunk 2]
```diff
@@ -75,40 +42,37 @@ const AddToken: React.FC<Props> = props => {
     setResultsData(results);
   }, [results]);
 
-  if (resultsData) return <ConfirmContainer />;
+  if (resultsData)
+    return (
+      <ConfirmContainer
```
Comment (andreyvEze) [inline client/src/pages/AddTokenPage/AddTokenPage.tsx:47]: What is this for? could we use Results component?

File: client/src/pages/EditTokenPage/EditTokenPage.tsx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+// todo: remove all any types
+const EditTokenPage: React.FC<any> = props => {
+  const { currentFioAddress, loading } = props;
+  const { name, publicAddresses, remaining } = currentFioAddress;
+
+  const [pubAddressesArr, changePubAddresses] = useState([]);
+  const [bundleCost, changeBundleCost] = useState(0);
+
+  // @ts-ignore
+  // eslint-disable-next-line no-unused-vars
```
Comment (andreyvEze) [inline client/src/pages/EditTokenPage/EditTokenPage.tsx:23]: remove

File: client/src/pages/EditTokenPage/components/PublicAddressEdit.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    handleClick(id, editedPubAddress);
+  };
+
+  const inputComponent = (
+    <input
+      type="text"
+      value={editedPubAddress}
+      onChange={onInputChange}
+      className={classes.input}
+      size={editedPubAddress.length * 1.2} // makes width of input almost the same as text
```
Comment (andreyvEze) [inline client/src/pages/EditTokenPage/components/PublicAddressEdit.tsx:58]: should we create const for `1.2`?

**Review by andreyvEze — CHANGES_REQUESTED** [review]
check comments

File: client/src/api/fio.ts [hunk 5]
```diff
@@ -452,4 +408,36 @@ export default class Fio {
...
+    keys: WalletKeys,
+    action: string,
+    params: any,
+  ): Promise<any> => {
+    this.setWalletFioSdk(keys);
+
+    if (!params.maxFee) params.maxFee = defaultFee;
+
+    try {
+      return await this.walletFioSDK.genericAction(action, params);
```
Comment (andreyvEze) [inline client/src/api/fio.ts:436]: is it not catch an error without `await`?

File: client/src/api/fio.ts [hunk 6]
```diff
@@ -452,4 +393,45 @@ export default class Fio {
...
+    } catch (err) {
+      this.logError(err);
+    } finally {
+      this.clearWalletFioSdk();
+    }
+  };
+
+  linkTokens = async (params: {
```
Comment (andreyvEze) [inline client/src/api/fio.ts:429]: remove

File: client/src/components/Badges/TokenBadge/TokenBadge.module.scss [hunk 7]
```diff
@@ -23,8 +23,14 @@
 }
 
 .addressContainer {
+  width: 100%;
   display: grid;
   grid-template-columns: 50px 150px auto;
+  align-items: center;
+
+  .publicAddressItem {
```
Comment (andreyvEze) [inline client/src/components/Badges/TokenBadge/TokenBadge.module.scss:31]: should we name it in common way? not connected to business logic

File: client/src/components/Badges/TokenBadge/TokenBadge.module.scss [hunk 8]
```diff
@@ -50,40 +56,50 @@
...
+        }
       }
     }
-  }
 
-  .pubAddress {
-    margin-top: 15px;
-    .title {
-      @include sf-pro-display-thin;
+    .pubAddress {
```
Comment (andreyvEze) [inline client/src/components/Badges/TokenBadge/TokenBadge.module.scss:86]: same here, rename

File: client/src/components/LinkTokenList/constants.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    buttonText: 'Edit',
+  },
+  [CONTAINER_NAMES.ADD]: {
+    title: 'Link your FIO Address',
+    buttonText: 'Link Now',
+  },
+};
+
+export const LOW_BALANCE_TEXT = {
+  buttonText: 'Renew',
```
Comment (andreyvEze) [inline client/src/components/LinkTokenList/constants.ts:23]: add bundled transactions not renew

File: client/src/pages/AddTokenPage/AddTokenPage.tsx [hunk 10]
```diff
@@ -1,114 +1,130 @@
...
-            <AddTokenForm
-              values={values}
-              push={push}
-              changeBundleCost={changeBundleCost}
-            />
-          </form>
+        render={formProps => (
+          <AddTokenForm
+            formProps={formProps}
+            containerName={CONTAINER_NAMES.ADD}
```
Comment (andreyvEze) [inline client/src/pages/AddTokenPage/AddTokenPage.tsx:115]: no need to set it here

File: client/src/pages/AddTokenPage/AddTokenPage.tsx [hunk 11]
```diff
@@ -1,114 +1,130 @@
...
+            results={resultsData}
+            name={name}
+            remaining={remaining}
```
Comment (andreyvEze) [inline client/src/pages/AddTokenPage/AddTokenPage.tsx:118]: could we pass just `currentFioAddress` not every prop in it?

File: client/src/pages/AddTokenPage/index.ts [hunk 12]
```diff
@@ -11,6 +12,8 @@ export default compose(
     createStructuredSelector({
       currentFioAddress,
     }),
-    {},
+    {
+      getFioAddresses,
```
Comment (andreyvEze) [inline client/src/pages/AddTokenPage/index.ts:16]: do we use it somewhere?

File: client/src/pages/EditTokenPage/EditTokenPage.tsx [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+          </h5>
+        </div>
+        {pubAddressesArr &&
+          pubAddressesArr.map(pubAddress => (
+            <PublicAddressEdit
+              {...pubAddress}
+              handleClick={handleEditTokenItem}
+              hasLowBalance={hasLowBalance}
+              key={pubAddress.id}
+            />
```
Comment (andreyvEze) [inline client/src/pages/EditTokenPage/EditTokenPage.tsx:209]: could we use final form here?

File: client/src/pages/EditTokenPage/EditTokenPage.tsx [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        action={CONFIRM_PIN_ACTIONS.EDIT_TOKEN}
+        processing={processing}
+        setProcessing={setProcessing}
+        fioWalletEdgeId={edgeWalletId}
+      />
+      <ActionContainer
+        containerName={CONTAINER_NAMES.EDIT}
+        name={name}
+        bundleCost={bundleCost}
+        remaining={remaining}
```
Comment (andreyvEze) [inline client/src/pages/EditTokenPage/EditTokenPage.tsx:188]: rename to `remainingBundledTxs`

File: client/src/pages/LinkTokenPage/TokenListPage.tsx [hunk 15]
```diff
@@ -1,51 +1,59 @@
-import React, { useState } from 'react';
-import { Button } from 'react-bootstrap';
-import { Link } from 'react-router-dom';
+import React, { useState, useEffect } from 'react';
```
Comment (andreyvEze) [inline client/src/pages/LinkTokenPage/TokenListPage.tsx:1]: rename file or folder

File: client/src/redux/fio/reducer.ts [hunk 16]
```diff
@@ -291,4 +355,15 @@ export default combineReducers({
         return state;
     }
   },
+  showTokenListInfoBadge(state: boolean = true, action) {
```
Comment (andreyvEze) [inline client/src/redux/fio/reducer.ts:358]: why do we need this?

File: client/src/pages/LinkTokenPage/TokenListPage.tsx [hunk 17]
```diff
@@ -1,51 +1,59 @@
...
+
+  const onClose = () => toggleTokenListInfoBadge(false);
+
+  useEffect(() => {
+    // show info badge if only FIO linked
+    toggleShowBadge(
+      showTokenListInfoBadge &&
+        publicAddresses &&
+        publicAddresses.length === 1 &&
+        publicAddresses[0].chainCode.toLowerCase() ===
```
Comment (andreyvEze) [inline client/src/pages/LinkTokenPage/TokenListPage.tsx:53]: fix this, now there is no FIO in the list

File: client/src/types.d.ts [hunk 18]
```diff
@@ -148,6 +150,8 @@ export type FioNameItemProps = {
   isPublic?: number;
   walletPublicKey?: string;
   publicAddresses?: PublicAddressDoublet[];
+  more?: boolean;
```
Comment (andreyvEze) [inline client/src/types.d.ts:153]: remove

File: client/src/pages/DeleteTokenPage/index.ts [hunk 19]
```diff
@@ -12,7 +12,7 @@ const reduxConnect = connect(
     currentFioAddress,
   }),
   {
-    onSubmit: linkTokens,
+    getFioAddresses,
```
Comment (andreyvEze) [inline client/src/pages/DeleteTokenPage/index.ts:15]: remove

File: client/src/util/hooks.ts [hunk 20]
```diff
@@ -40,3 +43,31 @@ export function useNonActiveUserRedirect() {
...
+  const fetchPublicAddresses = (incOffset: number = 0) =>
+    dispatch(getAllFioPubAddresses(fioAddress, limit, incOffset));
+
+  useEffect(() => {
+    if (!fioAddress) return;
+    fetchPublicAddresses();
+  }, []);
+
+  useEffect(() => {
+    if (currentFioAddress.more) {
```
Comment (andreyvEze) [inline client/src/util/hooks.ts:67]: how `more` is could be here? please check and refactor (remove places where `more` could be set to fio addresses)

File: client/src/pages/LinkTokenPage/TokenListPage.tsx (hunk 15 above)
Comment (trukhilio) [reply client/src/pages/LinkTokenPage/TokenListPage.tsx]: I don't have this folder, I have `TokenListPage/TokenListPage.tsx`

**Pull request merged**